pkg/provenance/provenance.go

```diff
@@ -147,122 +148,436 @@ func (o ObjectLineage) GetVersions() string {
 	}
 	sort.Ints(s)
 	//get all versions, sort by version, make string array of them
-	versions := make([]string, 0)
+	specs := make([]Spec, 0)
 	for _, version := range s {
-		versions = append(versions, fmt.Sprint(version)) //cast int to string
+		specs = append(specs, o[version])
 	}
-	return "[" + strings.Join(versions, ", ") + "]\n"
+	//get all versions, sort by version, make string array of them
+	outputs := make([]string, 0)
+	for _, spec := range specs {
+		outputs = append(outputs, fmt.Sprintf("%s: Version %d", spec.Timestamp, spec.Version)) //cast int to string
+	}
+	return "[" + strings.Join(outputs, ", \n") + "]\n"
 }
 
 //what happens if I delete the object?
 //need to delete the ObjectFullProvenance for the object
 //add type of ObjectFullProvenance, postgreses for example
 func (o ObjectLineage) SpecHistory() string {
+	s := make([]int, 0)
+	for _, value := range o {
+		s = append(s, value.Version)
+	}
+	sort.Ints(s)
+	//get all versions, sort by version, make slice of specs
+	specs := make([]Spec, 0)
+	for _, version := range s {
+		specs = append(specs, o[version]) //cast Spec to String
+	}
+
+	specStrings := make([]string, 0)
+	for _, spec := range specs {
+		specStrings = append(specStrings, spec.String())
+	}
+	return strings.Join(specStrings, "\n")
+}
+
+func (o ObjectLineage) SpecHistoryInterval(vNumStart, vNumEnd int) string {
+	//order keys so we append in order later, reference: https://blog.golang.org/go-maps-in-action#TOC_7.
 	s := make([]int, 0)
 	for _, value := range o {
 		s = append(s, value.Version)
 	}
 	sort.Ints(s)
 	//get all versions, sort by version, make string array of them
-	specs := make([]string, 0)
+	specs := make([]Spec, 0)
 	for _, version := range s {
-		specs = append(specs, fmt.Sprint(o[version])) //cast Spec to String
+		specs = append(specs, o[version]) //cast Spec to String
 	}
-	return strings.Join(specs, "\n")
+	specStrings := make([]string, 0)
+	for _, spec := range specs {
+		specStrings = append(specStrings, spec.String())
+	}
+	return strings.Join(specStrings, "\n")
 }
-func (o ObjectLineage) Bisect(field, value string) string {
-	s := make([]Spec, 0)
-	for _, spec := range o {
-		s = append(s, spec)
-	}
-	sort.Slice(s, func(i, j int) bool {
-		return s[i].Version < s[i].Version
-	})
-	if len(s) == 1 {
-		//check
-		if s[0].AttributeToData[field] == value {
-			return strconv.Itoa(1)
+func BuildAttributeRelationships(specs []Spec, allQueryPairs [][]string) map[string][][]string{
+	mapRelationships := make(map[string][][]string, 0)
+	// a map from top level attribute to array of pairs (represented as 2 len array)
+	// mapRelationships with one top level object users, looks like This:
+	//map[users:[[username pallavi] [password pass123]]]
+	//after I do
+	for _, spec := range specs {
```

Comment:
I am wondering why we need to iterate over all the specs? Will it not be sufficient to just use a single Spec object  and use its AttributeToData map to build the relationships?

Reply:
hmm good question, initially I wasn't sure if each spec fully represents the map relationship, like for example, i was thinking that some attributes may be omitted if they are not set or specified, but perhaps a later kubectl apply -f  could change that in a following version   . 
So perhaps no users are specified initially, then the map for users in spec 1 is empty? Then the maprelationship may not be built correctly, if the versions 2-5 are not taken into account.

I don't know, do you think this is a concern? I can test it later but yes I agree that iterating over all the specs is likely unnecessary